rules.py

```diff
@@ -911,6 +911,26 @@ def check(self):
             raise base.ElectionTreeError(
                 "The Election File has following issues with the identifiers.", error_log)
 
+class CandidatesMissingPartyData(base.TreeRule):
+    """Each Candidate should have party data associated with them.
+
+    A Candidate object that has no PartyId attached to them should be picked up
+    within this class and returned to the user as a warning."""
+
+    def check(self):
+        error_ids = []
+        party_ids = self.get_elements_by_class(self.election_tree, "Party")
```

Comment:
You shouldn't have to worry about getting the set of party IDs, since if a party ID is invalid that will trigger a schema error. You can just go through each `Candidate` object and make sure its `PartyId` element is present and non-empty.